node/rate_limiter.js

```diff
@@ -0,0 +1,168 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+var DEFAULT_SERVICE_RPS_LIMIT = 5000;
+var DEFAULT_TOTAL_RPS_LIMIT = 20000;
+var DEFAULT_BUCKETS = 20;
+
+function RateLimiter(options) {
+    if (!(this instanceof RateLimiter)) {
+        return new RateLimiter(options);
+    }
+    var self = this;
+
+    self.timers = options.timers;
+
+    self.buckets = options.buckets || DEFAULT_BUCKETS;
+    assert(self.buckets > 0 && self.buckets <= 1000, 'counter buckets should between (0 1000]');
+
+    self.defaultServiceRpsLimit = options.defaultServiceRpsLimit || DEFAULT_SERVICE_RPS_LIMIT;
+    self.totalRpsLimit = options.totalRpsLimit || DEFAULT_TOTAL_RPS_LIMIT;
+    self.rpsLimitForServiceName = options.rpsLimitForServiceName || Object.create(null);
+    self.counters = Object.create(null);
+    self.totalRequestCounter = self.createCounter(self.totalRpsLimit);
+
+    self.refreshDelay = 1000 / self.buckets;
+    self.refresh();
+
+    self.destroyed = false;
+}
+
+RateLimiter.prototype.type = 'tchannel.hyperbahn.rate-limiting';
+
+RateLimiter.prototype.refresh =
+function refresh() {
+    var self = this;
+
+    self.udpateCounter(self.totalRequestCounter);
+
+    var serviceNames = Object.keys(self.counters);
+    for (var i = 0; i < serviceNames.length; i++) {
+        self.udpateCounter(self.counters[serviceNames[i]]);
+    }
+
+    self.refreshTimer = self.timers.setTimeout(
+        function refresh() {
+            self.refresh();
+        },
+        self.refreshDelay
+    );
+};
+
+RateLimiter.prototype.udpateCounter =
+function udpateCounter(counter) {
+    var self = this;
+
+    // update the sliding window
+    var next = (counter.index + 1) % self.buckets;
+    if (counter.buckets[next]) {
+        // offset the bucket being moved out
+        counter.rps -= counter.buckets[next];
+    }
+
+    assert(counter.rps >= 0, 'rps should always be larger equal to 0');
+    counter.index = next;
+    counter.buckets[counter.index] = 0;
+};
+
+RateLimiter.prototype.createCounter =
+function createCounter(rpsLimit) {
+    var self = this;
+
+    var counter = Object.create(null);
```

Comment:
As an aside; do not use `Object.create(null)` for anything other then dictionaries